### PR TITLE
Uber PR

### PR DESCRIPTION
**Pull request by andy128k**
- Rewrite async API to use promises
- Remove sync API
- Migrate test suite to Jest
- Switch from JSHint to ESLint

Closes https://github.com/lydell/source-map-resolve/pull/16
Closes https://github.com/lydell/source-map-resolve/pull/17
Closes https://github.com/lydell/source-map-resolve/pull/18
Closes https://github.com/lydell/source-map-resolve/pull/19

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -5,10 +5,11 @@ module.exports = {
     "es6": true,
     "node": true
   },
-  "plugins": ["jest"],
+  "plugins": ["jest", "prettier"],
```
Comment (lydell) [inline .eslintrc.js:8]: Can you remove eslint-plugin-prettier, please? See https://github.com/prettier/prettier/issues/7828

File: .jshintrc [hunk 2]
```diff
@@ -3,7 +3,8 @@
   "camelcase": true,
   "curly": false,
   "eqeqeq": true,
-  "es3": true,
+  "es3": false,
```
Comment (lydell) [inline .jshintrc:6]: Can you remove this file?

File: package.json [hunk 3]
```diff
@@ -22,16 +22,25 @@
     "index.js"
   ],
   "scripts": {
-    "lint": "jshint index.js test/",
-    "unit": "node test/index.js && node test/read.js && node test/windows.js",
+    "fmt": "prettier --write index.js test/",
+    "lint": "eslint index.js test/",
+    "unit": "jest",
     "test": "npm run lint && npm run unit"
```
Comment (lydell) [inline package.json:28]: Can you use this approach?

https://github.com/lydell/js-tokens/blob/417dc4085a7e3468d01a6371abc81373188dbc43/package.json#L5-L9

It requires adding stuff to .prettierignore and .eslintignore as needed.

File: .eslintrc.js (hunk 1 above)
Comment (lydell) [reply .eslintrc.js:8]: Also, this ESLint config is kinda ok: https://github.com/lydell/js-tokens/blob/9992be622597981b00d711481036c01687300b3d/.eslintrc.js

It automatically enforces som ES6 syntax.

Comment (andy128k) [reply .eslintrc.js]: I took it but slightly adjusted.

`prefer-destructuring` for variable assignment is *ugly*, IMHO.
Also `jest/no-try-expect` is relaxed to `warn`. Jest's `expect(..).toThrow(..)` doesn't allow to make multiple matches (or I do not know how to do it :) )

File: .jshintrc (hunk 2 above)
Comment (andy128k) [reply .jshintrc]: done

File: package.json (hunk 3 above)
Comment (andy128k) [reply package.json]: Do you prefer blacklist files instead of whitelisting what needs to be linted/formatted?

Comment (lydell) [reply package.json]: Yes!

**Comment by lydell** [comment]
Sorry, I pressed the "Update branch" button but it doesn’t seem to be working. I wanted GitHub Actions to appear here.

**Comment by andy128k** [comment]
So, you need this https://github.com/lydell/source-map-resolve/pull/22

**Comment by andy128k** [comment]
@lydell I also updated readme.md. Please review.

**Comment by andy128k** [comment]
@lydell Is anything missing in this PR? I mean, actions run and pass.

**Comment by lydell** [comment]
Sorry, I haven’t had time to review. But it’s on my list.

**Comment by lydell** [comment]
Truly awesome! :medal_sports: 

Sorry for being so slow to review. I’m merging this into the `next` branch. I want to do some more tweaks before going into `master` (over the next couple of weeks).

Feel free to make more improvements if you like!

**Pull request merged**